### PR TITLE
Add admin oc apply for geneva actions

### PR DESCRIPTION
**Pull request by asalkeld**
geneva side: https://msazure.visualstudio.com/One/_git/Compute-ARO-GenevaActions/pullrequest/2788499?_a=overview

### CODE REVIEW

File: pkg/frontend/frontend.go [hunk 1]
```diff
@@ -178,6 +178,12 @@ func (f *frontend) authenticatedRoutes(r *mux.Router) {
 
 	s.Methods(http.MethodGet).HandlerFunc(f.getAdminKubernetesObjects).Name("getAdminKubernetesObjects")
 
+	s = r.
+		Path("/admin/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/{resourceProviderNamespace}/{resourceType}/{resourceName}/kubernetesobjects").
+		Subrouter()
+
```
Comment (jim-minter) [inline pkg/frontend/frontend.go:184]:
```suggestion
```

(reuse above subrouter)

File: pkg/frontend/kubeactions/kubeactions.go [hunk 2]
```diff
@@ -160,6 +163,63 @@ func (ka *kubeactions) List(ctx context.Context, oc *api.OpenShiftCluster, kind,
 	return ul.MarshalJSON()
 }
 
+func (ka *kubeactions) applyOne(ctx context.Context, dyn dynamic.Interface, grs []*restmapper.APIGroupResources, un *unstructured.Unstructured) error {
+	if strings.EqualFold(un.GetKind(), "secret") { // TODO do we need this?
```
Comment (jim-minter) [inline pkg/frontend/kubeactions/kubeactions.go:167]: yes, we do.

File: pkg/frontend/kubeactions/kubeactions.go [hunk 3]
```diff
@@ -160,6 +163,63 @@ func (ka *kubeactions) List(ctx context.Context, oc *api.OpenShiftCluster, kind,
...
+		return fmt.Errorf("access to secrets is forbidden")
+	}
+	gvrs := ka.findGVR(grs, un.GetKind())
+	if len(gvrs) == 0 {
+		return fmt.Errorf("Kind %s not found", un.GetKind())
+	}
+	if len(gvrs) > 1 {
+		return fmt.Errorf("Kind %s did not uniquely match one GRV", un.GetKind())
```
Comment (jim-minter) [inline pkg/frontend/kubeactions/kubeactions.go:175]: internal errors start lower case so that they can be composed.  GRV is misspelled.  Reuse the error message from Get().

File: pkg/frontend/kubeactions/kubeactions.go [hunk 4]
```diff
@@ -160,6 +163,63 @@ func (ka *kubeactions) List(ctx context.Context, oc *api.OpenShiftCluster, kind,
...
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		_obj, err := dyn.Resource(*gvr).Namespace(un.GetNamespace()).Get(un.GetName(), metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		un.SetResourceVersion(_obj.GetResourceVersion())
+		_, err = dyn.Resource(*gvr).Namespace(un.GetNamespace()).Update(un, metav1.UpdateOptions{})
+		return err
+	})
```
Comment (jim-minter) [inline pkg/frontend/kubeactions/kubeactions.go:195]: a rare occasion where RetryOnConflict should not be used.  Try the update once, and if it fails, return the error and let the user decide what to do.  Otherwise the following could happen:

1. SRE gets object
2. Cluster user updates object
3. SRE updates object, losing changes from 2

File: pkg/frontend/kubeactions/kubeactions.go [hunk 5]
```diff
@@ -160,6 +163,63 @@ func (ka *kubeactions) List(ctx context.Context, oc *api.OpenShiftCluster, kind,
...
+	}
+	if len(gvrs) > 1 {
+		return fmt.Errorf("Kind %s did not uniquely match one GRV", un.GetKind())
+	}
+	gvr := gvrs[0]
+
+	ka.log.Debugf("appling kube object %s %s %s ", un.GetKind(), un.GetNamespace(), un.GetName())
+
+	_, err := dyn.Resource(*gvr).Namespace(un.GetNamespace()).Create(un, metav1.CreateOptions{})
+	if err == nil || !errors.IsAlreadyExists(err) {
```
Comment (jim-minter) [inline pkg/frontend/kubeactions/kubeactions.go:182]:
```suggestion
	if !errors.IsAlreadyExists(err) {
```

File: pkg/frontend/kubeactions/kubeactions.go [hunk 6]
```diff
@@ -160,6 +163,63 @@ func (ka *kubeactions) List(ctx context.Context, oc *api.OpenShiftCluster, kind,
...
+	gvrs := ka.findGVR(grs, un.GetKind())
+	if len(gvrs) == 0 {
+		return fmt.Errorf("Kind %s not found", un.GetKind())
+	}
+	if len(gvrs) > 1 {
+		return fmt.Errorf("Kind %s did not uniquely match one GRV", un.GetKind())
+	}
+	gvr := gvrs[0]
+
+	ka.log.Debugf("appling kube object %s %s %s ", un.GetKind(), un.GetNamespace(), un.GetName())
```
Comment (jim-minter) [inline pkg/frontend/kubeactions/kubeactions.go:179]: spelling

File: pkg/frontend/admin_openshiftcluster_kubernetesobjects_post.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	"github.com/Azure/ARO-RP/pkg/frontend/middleware"
+)
+
+func (f *frontend) postAdminKubernetesObjects(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	log := ctx.Value(middleware.ContextKeyLog).(*logrus.Entry)
+	r.URL.Path = filepath.Dir(r.URL.Path)
+
+	body := r.Context().Value(middleware.ContextKeyBody).([]byte)
+	if len(body) > 0 && !json.Valid(body) {
```
Comment (jim-minter) [inline pkg/frontend/admin_openshiftcluster_kubernetesobjects_post.go:27]:
```suggestion
	if len(body) == 0 || !json.Valid(body) {
```

File: pkg/frontend/kubeactions/kubeactions.go [hunk 8]
```diff
@@ -27,6 +29,7 @@ import (
 type Interface interface {
 	Get(ctx context.Context, oc *api.OpenShiftCluster, kind, namespace, name string) ([]byte, error)
 	List(ctx context.Context, oc *api.OpenShiftCluster, kind, namespace string) ([]byte, error)
+	Apply(ctx context.Context, oc *api.OpenShiftCluster, body []byte) error
```
Comment (jim-minter) [inline pkg/frontend/kubeactions/kubeactions.go:32]: I think that CreateOrUpdate would be a clearer name throughout than Apply.

File: pkg/frontend/kubeactions/kubeactions.go [hunk 9]
```diff
@@ -160,6 +163,63 @@ func (ka *kubeactions) List(ctx context.Context, oc *api.OpenShiftCluster, kind,
...
+	err = objs.UnmarshalJSON(body)
+	if err != nil {
+		return err
+	}
+	for objIx := range objs.Items {
+		err = ka.applyOne(ctx, dyn, grs, &objs.Items[objIx])
+		if err != nil {
+			return err
+		}
+	}
```
Comment (jim-minter) [inline pkg/frontend/kubeactions/kubeactions.go:219]: let's not support this for now, I think it's too powerful and the error handling is too complex.

File: pkg/frontend/kubeactions/kubeactions.go (hunk 6 above)
Comment (jim-minter) [inline pkg/frontend/kubeactions/kubeactions.go:179]: I think we should log the current and the new state of the object.

File: pkg/frontend/kubeactions/kubeactions.go (hunk 8 above)
Comment (jim-minter) [reply pkg/frontend/kubeactions/kubeactions.go:32]: Please do a delete operation as a follow-up PR.

File: pkg/frontend/kubeactions/kubeactions.go (hunk 6 above)
Comment (asalkeld) [reply pkg/frontend/kubeactions/kubeactions.go:179]: when you say "state" do you mean the whole object? If so I'd worry about secrets being in the env section. I'll have a look and see what sync pod does..

File: pkg/frontend/kubeactions/kubeactions.go [hunk 10]
```diff
@@ -162,6 +166,61 @@ func (ka *kubeactions) List(ctx context.Context, oc *api.OpenShiftCluster, kind,
 	return ul.MarshalJSON()
 }
 
+func printDiff(log *logrus.Entry, existing, o *unstructured.Unstructured) {
+	gk := o.GroupVersionKind().GroupKind()
+	if gk.String() != "Secret" { // shouldn't be needed, but just in case
```
Comment (nilsanderselde) [inline pkg/frontend/kubeactions/kubeactions.go:171]: `gk` is only used once - you could just do
```
if o.GroupVersionKind().GroupKind().String() != "Secret" {
```

File: pkg/frontend/admin_openshiftcluster_kubernetesobjects.go [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+	body := r.Context().Value(middleware.ContextKeyBody).([]byte)
+	if len(body) == 0 || !json.Valid(body) {
+		api.WriteError(w, http.StatusBadRequest, api.CloudErrorCodeInvalidRequestContent, "", "The request content was invalid and could not be deserialized.")
+		return
+	}
+
+	err := f._postAdminKubernetesObjects(ctx, r)
+
+	reply(log, w, nil, nil, err)
```
Comment (asalkeld) [inline pkg/frontend/admin_openshiftcluster_kubernetesobjects.go:105]: this should be adminReply

File: pkg/frontend/admin_openshiftcluster_kubernetesobjects.go [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+// prevent mischief
+var rxKubernetesString = regexp.MustCompile(`(?i)^[-a-z0-9]{0,255}$`)
+
+func validateAdminKubernetesObjectsNonCustomer(method, kind, namespace, name string) error {
+	if namespace != "" &&
+		namespace != "default" &&
+		namespace != "openshift" &&
+		!strings.HasPrefix(string(namespace), "kube-") &&
+		!strings.HasPrefix(string(namespace), "openshift-") {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, "", "The provided namespace '%s' is invalid.", namespace)
```
Comment (asalkeld) [inline pkg/frontend/admin_openshiftcluster_kubernetesobjects.go:146]: Should really say access denied, as they are not invalid

File: pkg/frontend/admin_openshiftcluster_kubernetesobjects.go [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+// rxKubernetesString is weaker than Kubernetes validation, but strong enough to
+// prevent mischief
+var rxKubernetesString = regexp.MustCompile(`(?i)^[-a-z0-9]{0,255}$`)
+
+func validateAdminKubernetesObjectsNonCustomer(method, kind, namespace, name string) error {
+	if namespace != "" &&
+		namespace != "default" &&
+		namespace != "openshift" &&
+		!strings.HasPrefix(string(namespace), "kube-") &&
+		!strings.HasPrefix(string(namespace), "openshift-") {
```
Comment (nilsanderselde) [inline pkg/frontend/admin_openshiftcluster_kubernetesobjects.go:145]: It might just be me, but I think the following is much easier to read:

```
	if !(
		namespace == "" ||
		namespace == "default" ||
		namespace == "openshift" ||
		strings.HasPrefix(string(namespace), "kube-") ||
		strings.HasPrefix(string(namespace), "openshift-")
	) {
```

In plain English, it's the different between "not A and not B and not C and not D" and "neither A, B, C nor D".

File: pkg/frontend/admin_openshiftcluster_kubernetesobjects.go (hunk 12 above)
Comment (asalkeld) [reply pkg/frontend/admin_openshiftcluster_kubernetesobjects.go:146]: I wonder if this check should be in the .net client, as that is faster to update if in a pinch we need to fine tune this..

File: pkg/frontend/admin_openshiftcluster_kubernetesobjects.go (hunk 13 above)
Comment (jim-minter) [reply pkg/frontend/admin_openshiftcluster_kubernetesobjects.go:145]: cc De Morgan.

this function is slated to go to a helper in a future PR.

File: pkg/frontend/admin_openshiftcluster_kubernetesobjects.go (hunk 12 above)
Comment (jim-minter) [reply pkg/frontend/admin_openshiftcluster_kubernetesobjects.go]: I'm happy with it here. Updated.

**Comment by jim-minter** [comment]
thanks for the additional :eyes: , will merge on green
